### PR TITLE
[CHAT-1306]: Init

### PR DESCRIPTION
**Pull request by TsvetanMilanov**
# Purpose

Create package which follows the documentation of AWS -> https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-using-tokens-verifying-a-jwt.html

### CODE REVIEW

File: lib/validator.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+export class JwtValidator {
+    private validationOptions: IValidationOptions;
+    private cacheKeys?: IMapOfKidToPublicKey;
+
+    constructor(
+        private config: IJwtValidatorConfig,
+        validationOptions?: IValidationOptions,
+    ) {
+        this.validationOptions = Object.assign({}, DefaultValidationOptions, validationOptions || {});
```
Comment (Mitko-Kerezov) [inline lib/validator.ts:26]: Defaulting to `{}` is unnecessary

```suggestion
        this.validationOptions = Object.assign({}, DefaultValidationOptions, validationOptions);
```

File: lib/validator.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    public async validateToken<T>(token: string): Promise<IValidateTokenResult<T>> {
+        const result: IValidateTokenResult<T> = { isValid: false };
+        try {
+            const tokenSections = (token || "").split(".");
+            if (tokenSections.length < 2) {
```
Comment (Mitko-Kerezov) [inline lib/validator.ts:33]: According to docs, shouldn't this be:
```suggestion
            if (tokenSections.length !== 3) {
```

File: lib/validator.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        } catch (err) {
+            result.error = err;
+        }
+
+        return result;
+    }
+
+    private async getPublicKeys(): Promise<IMapOfKidToPublicKey> {
+        if (!this.cacheKeys) {
+            const separator = this.config.issuerUrl.endsWith("/") ? "" : "/";
```
Comment (Mitko-Kerezov) [inline lib/validator.ts:74]: `"/"` could be extracted into a local constant for this file

File: lib/validator.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            if (this.validationOptions.validateExp && (currentSeconds > claim.exp || currentSeconds < claim.auth_time)) {
+                throw new Error("claim is expired or invalid");
+            }
+
+            if (this.validationOptions.validateIss && claim.iss !== this.config.issuerUrl) {
+                throw new Error("claim issuer is invalid");
+            }
+
+            if (this.validationOptions.validateAud && claim.aud !== this.config.clientId) {
+                throw new Error("claim aud is invalid");
```
Comment (Mitko-Kerezov) [inline lib/validator.ts:56]:
```suggestion
                throw new Error("claim audience is invalid");
```

File: spec/index.spec.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    iss: "https://nchat.us.auth0.com/",
+    sub: "auth0|5f5c9adf1d80b10078e5dccb",
+    aud: "AqG0WVSYf7x33jXEd0g3nLrjWctvrv7Y",
+    iat: 1599917619,
+    exp: 1599953619,
+    nonce: "0bNqna.PoXj90ZdaCdY3OygkEr7bJfVGE5vunp_-_oh",
+};
+
+const getDefaultConfig = (): IJwtValidatorConfig => {
+    const config: IJwtValidatorConfig = { issuerUrl: "https://nchat.us.auth0.com/", clientId: "AqG0WVSYf7x33jXEd0g3nLrjWctvrv7Y", tokenUse: "access" };
```
Comment (Mitko-Kerezov) [inline spec/index.spec.ts:22]:
```suggestion
    const config: IJwtValidatorConfig = { issuerUrl: payload.iss, clientId: payload.aud, tokenUse: "access" };
```

File: spec/index.spec.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+            const result = await validator.validateToken(token);
+
+            expect(result.isValid).toBeTrue();
+            expect(result.error).toBeUndefined();
+            expect(result.claim).toEqual(payload as any);
+        });
+
+        it("should validate iss.", async () => {
+            const config = getDefaultConfig();
+            config.issuerUrl = "https://nchat.us.auth0.com";
```
Comment (Mitko-Kerezov) [inline spec/index.spec.ts:47]:
```suggestion
            config.issuerUrl = config.issuerUrl.slice(0, -1);
```

File: spec/index.spec.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+};
+
+const expectError = (result: IValidateTokenResult<IEmailClaim>, errorMessage: string) => {
+    expect(result.isValid).toBeFalse();
+    expect(result.error).toEqual(new Error(errorMessage));
+    expect(result.claim).toBeUndefined();
+};
+
+describe("JwtValidator", () => {
+    describe("validateToken", () => {
```
Comment (Mitko-Kerezov) [inline spec/index.spec.ts:34]: you should add a test for `"requested token is invalid"` error

**Review by Mitko-Kerezov — CHANGES_REQUESTED** [review]

File: lib/validator.ts (hunk 2 above)
Comment (TsvetanMilanov) [reply lib/validator.ts:33]: Yes, but in their example code is `< 2` 😄

**Review by Mitko-Kerezov — APPROVED** [review]

**Pull request merged**